pallets/gear-scheduler/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+    #[pallet::storage_version(SCHEDULER_STORAGE_VERSION)]
+    pub struct Pallet<T>(_);
+
+    // Gear Scheduler Pallet error type.
+    //
+    // Used as inner error type for `Scheduler` implementation.
+    #[pallet::error]
+    pub enum Error<T> {
+        /// Occurs when given task already exists in tasks scope.
+        DuplicateTask,
+        /// Occurs when task wasn't found in storage.
+        TaskNotFound,
+    }
+
+    // Implementation of `DequeueError` for `Error<T>`
+    // usage as `Queue::Error`.
+    impl<T: crate::Config> TasksScopeError for Error<T> {
+        fn duplicate_task() -> Self {
+            Self::DuplicateTask
+        }
+
+        fn task_not_found() -> Self {
+            Self::TaskNotFound
+        }
+    }
+
+    /// Account Id type of the task.
+    type AccountId<T> = <T as frame_system::Config>::AccountId;
+
+    /// Task type of the scheduler.
+    type Task<T> = ScheduledTask<AccountId<T>>;
+
+    // BTreeSet used to exclude duplicates and always keep collection sorted.
+    /// Missed blocks collection type.
+    ///
+    /// Defines block number, which should already contain no tasks,
+    /// because they were processed before.
+    /// Missed blocks processing prioritized.
+    type MissedBlocksCollection<T> = BTreeSet<BlockNumberFor<T>>;
+
+    // Below goes storages and their gear's wrapper implementations.
+    //
+    // Note, that we declare storages private to avoid outside
+    // interaction with them, but wrappers - public to be able
+    // use them as generic parameters in public `Scheduler`
+    // implementation.
+
+    // ----
+
+    // Private storage for missed blocks collection.
+    #[pallet::storage]
+    type MissedBlocks<T> = StorageValue<_, MissedBlocksCollection<T>>;
+
+    // Public wrap of the missed blocks collection.
+    common::wrap_storage_value!(
+        storage: MissedBlocks,
+        name: MissedBlocksWrap,
+        value: MissedBlocksCollection<T>
+    );
+
+    // ----
+
+    // Private storage for tasks scope elements.
+    // Primary item stored as second key of double map for optimization.
+    // Value here is useless, so unit type used as space saver:
+    // `assert_eq!(().encode().len(), 0)`
+    #[pallet::storage]
+    type TasksScope<T: Config> =
+        StorageDoubleMap<_, Identity, BlockNumberFor<T>, Identity, Task<T>, ()>;
+
+    // Public wrap of the mailbox elements.
+    common::wrap_extended_storage_double_map!(
+        storage: TasksScope,
+        name: TasksScopeWrap,
+        key1: BlockNumberFor<T>,
+        key2: Task<T>,
+        value: (),
+        length: usize
+    );
+
+    // ----
+
+    // Below goes callbacks, used for task scope algorithm.
+    //
+    // Note, that they are public like storage wrappers
+    // only to be able use as public trait's generics.
+
+    // ----
+
+    /// Callback function for success `add` and `delete` actions.
+    pub struct OnAffect<T: crate::Config>(PhantomData<T>);
```

Comment:
`OnAffect` sounds a bit over the top to me. Isn't the conventional `OnChange` more appropriate here?